### PR TITLE
 Move protoc generation to binary crate (#5718)

### LINKED ISSUE #5718: Move protoc generation to binary crate
### Is your feature request related to a problem or challenge?

As  of https://github.com/apache/arrow-datafusion/pull/3950 DataFusion vendors the generated protobuf code, for a variety of reasons linked to from the PR. Unfortunately in order to avoid drift this needs to pin the version of tonic-build, etc... used to generate the code. The current approach of using a build.rs causes this to leak into downstreams

### Describe the solution you'd like

Move the protoc generation into a non-published subcrate as was recently done in arrow-rs - https://github.com/apache/arrow-rs/pull/3927

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by l0kr**
# Which issue does this PR close?

Closes #5718 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Could we get a CI check, similar to https://github.com/apache/arrow-rs/pull/3927/files#diff-7a5f24a0bed6232c2297eba93812cdaeb26327ba955eabc0a04effc21e95d52bR65

File: datafusion/proto/gen/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# or more contributor license agreements.  See the NOTICE file
```
Comment (tustvold) [inline datafusion/proto/gen/Cargo.toml:1]: This license header appears to be truncated?

**Comment by l0kr** [comment]
> Could we get a CI check, similar to https://github.com/apache/arrow-rs/pull/3927/files#diff-7a5f24a0bed6232c2297eba93812cdaeb26327ba955eabc0a04effc21e95d52bR65

Should I create another yaml or add check to existing one?

**Comment by tustvold** [comment]
I think adding it to https://github.com/apache/arrow-datafusion/blob/main/.github/workflows/rust.yml would make sense

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -320,6 +320,20 @@ jobs:
...
+    runs-on: ubuntu-latest
+    container:
+      image: amd64/rust
+    steps:
+      - uses: actions/checkout@v3
+      - name: Setup Rust toolchain
+        uses: ./.github/actions/setup-builder
+      - name: Run gen
+        run: ./regen.sh
+        working-directory: ./datafusion/proto
```
Comment (tustvold) [inline .github/workflows/rust.yml:334]: This part shouldn't be necessary, as the script already handles this, but I suppose it can't hurt

**Review by tustvold — APPROVED** [review]

File: datafusion/proto/Cargo.toml [hunk 3]
```diff
@@ -54,8 +54,3 @@ serde_json = { version = "1.0", optional = true }
 [dev-dependencies]
 doc-comment = "0.3"
 tokio = "1.18"
-
-[build-dependencies]
-# Pin these dependencies so that the generated output is deterministic
```
Comment (alamb) [inline datafusion/proto/Cargo.toml:59]: 🎉

**Comment by alamb** [comment]
I verified this works great -- https://github.com/apache/arrow-datafusion/pull/5760#issue-1642777613 

Thanks @tustvold  and @l0kr

**Pull request merged**